### ISSUE TITLE
`traverse` can be slow

### ISSUE THREAD

**Issue by bovee** [onecodex/taxonomy#4]
Running e.g. `prune_to` on a large-ish taxonomy (>100k nodes) takes minutes mostly because the `traverse` method takes a long time. I think this is mostly because we're doing a ton of cloning in this method and malloc is taking a lot of time.

<img width="1170" alt="Screen Shot 2019-08-14 at 4 14 06 PM" src="https://user-images.githubusercontent.com/1778539/63062765-de057f80-beae-11e9-8f8a-b356266d52b0.png">

**Comment by bovee** [comment]
I tried writing a `impl<'t> Iterator for TaxonomyIterator<'t, usize, f32> ...` specialized method, but I keep getting `conflicting implementation for 'taxonomy::TaxonomyIterator<'_, usize, f32>'` errors. I tried rearranging the `clones`, but I was also having issues with lifetimes with that.

**Comment by Keats** [comment]
Unrelated: what tool is the screenshot from?

**Comment by luizirber** [comment]
@Keats instruments, you can launch it from the command line:
https://help.apple.com/instruments/mac/10.0/#/devba105ecc

**Comment by Keats** [comment]
Thanks!
It does look like it's missing the number of allocations (temp or not) as well as the memory usage.

@bovee I have an Ubuntu VM with heaptrack set up. If you can get me a runnable example, I can send you the profiling for allocations

**Comment by bovee** [comment]
@Keats Thanks! There's a separate mode for allocations; it looks like it's making a quarter million or so transient small ones per second (not super adjusted to the interface to say where they are, but guessing they're copying keys around):

<img width="1024" alt="Screen Shot 2019-08-15 at 2 09 21 PM" src="https://user-images.githubusercontent.com/1778539/63127132-7a3e8d80-bf66-11e9-95b4-1cd33ed6d188.png">

I'm running a fairly simple example using one of our database `build.json`'s:
```python
import json
from taxonomy import Taxonomy
tax = Taxonomy.from_json(json.dumps(json.load(open('.../build.json'))['taxonomy']['node_link_data']))
_ = tax.prune(keep=['562'])  # this line basically kills the process
```

I think this may also be responsible for how hideously slow iterating over the taxonomy in Python is (although there may be other issues there with how I construct the Python iterator).

**Comment by Keats** [comment]
With debug=True in the release profile of Cargo.toml, heaptrack can tell you exactly which line is allocating, how many allocations and their total/peak values. Maybe instruments does the same, I'll try tomorrow with the proxy.
Can I get a build.json from s3?

**Comment by luizirber** [comment]
@keats s3://refgenomics-datafiles/dbs/mg_complete_20181218_extra/build.kmerdb.json

**Comment by luizirber** [comment]
hmm, but taxonomy doesn't work with the current mgo db?
```

**Comment by bovee** [comment]
@luizirber That error was fixed in b5794ad9e33065ceff8e4e73f7b3130089dc4c1c; once we get something for this issue we can deploy both fixes as 0.3.2 on pypi, etc.

**Comment by bovee** [comment]
Okay, this turned out to be a bug where `tax.children('1')` returned e.g. `['1', '2', '10239']` so we were infinitely recursing (every time we cleared the queue root was still there). The actual speed of the code is fine; it takes ~47 ms to prune the taxonomy now.
